edb/edgeql/compiler/policies.py

```diff
@@ -248,6 +266,100 @@ def get_rewrite_filter(
     return filter_expr
 
 
+def _always_references_set(
+    ir: irast.Set | irast.Expr,
+    ref: irast.Set,
+    inverted: bool=False,
+) -> bool:
+    """Return whether *ir* "always references" *ref*
+
+    The idea here is to check whether *ir* references *ref* in a way
+    that ensures that postgres will actually look at *ref* when
+    executing.
+
+    Fortunately postgres doesn't seem to do anything too crazy here(??),
+    so we mostly just have to understand how it works with boolean
+    operators, IF, and coalesce.
+    But we also need to be able to propagate it through other operations.
+
+    We need *ref* to be referenced in *every* conjunct of an AND.
+    We need it referenced by *at least one* disjunct of an OR.
+    But because of DeMorgan's law (which postgres understands),
+    OR sometimes needs to be treated like AND.
+
+    We track the invertedness and invert the AND and OR behavior when
+    underneath a NOT, kind of for fun.
+    """
+    if isinstance(ir, irast.Set):
+        if ir is ref:
+            return True
+        ir = ir.expr
+
+    match ir:
+        case irast.SelectStmt(result=result):
```

Comment:
Is it guaranteed that we would ever only see `select` in this context and not anything else?

Reply:
DML isn't allowed in access policies.

Reply:
I'd probably still add an assertion

Reply:
We return false for the fall-through case, though, so I feel fine about it